### PR TITLE
Annotated a few classes of the text package

### PR DESCRIPTION
**Pull request by Kakarot-SSJ4**
Completed the annotations for the following classes:

- ExtendedMessageFormat.java

- FormattableUtils.java

- StrMatcher.java

- StrTokenizer.java

- CharSequenceTranslator.java

- EntityArrays.java

- JavaUnicodeEscaper.java

- LookupTranslator.java

- OctalUnescaper.java

- UnicodeEscaper.java

Kindly ignore WordUtils.java, I have not yet annotated it, I am stuck in that class.

### CODE REVIEW

File: Stubs/Matcher.astub [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import org.checkerframework.common.value.qual.MinLen;
```
Comment (kelloggm) [inline Stubs/Matcher.astub]: We agreed in our meeting earlier that you should use the annotations in the JDK for the `Matcher` class, so this stub file shouldn't be necessary.

File: Stubs/Matcher.astub [hunk 2]
```diff
@@ -7,12 +8,12 @@ class Matcher implements MatchResult {
...
     Matcher reset();
     Matcher reset(CharSequence arg0);
-    int start();
-    int start(int arg0);
-    int start(String arg0);
-    int end();
-    int end(int arg0);
-    int end(String arg0);
+    @NonNegative int start();
+    @NonNegative int start(int arg0);
```
Comment (kelloggm) [inline Stubs/Matcher.astub]: In particular, this method should return `@GTENegativeOne int`, and only accept `@NonNegative int`

File: pom.xml [hunk 3]
```diff
@@ -934,7 +934,7 @@
         <arg>-Xbootclasspath/p:${annotatedJdk}</arg>
         <!-- -Awarns turns type-checking warnings into errors. -->
         <!-- <arg>-Awarns</arg> -->
-        <arg>-AonlyDefs=^org\.apache\.commons\.lang3\.builder\.</arg>
+        <arg>-AonlyDefs=^org\.apache\.commons\.lang3\.text\.</arg>
```
Comment (kelloggm) [inline pom.xml]: Before merging, you'll want to change this so that it adds to what's being checked, rather than replacing it. I'm leaving this comment here so that we remember.

File: src/main/java/org/apache/commons/lang3/text/ExtendedMessageFormat.java [hunk 4]
```diff
@@ -143,6 +147,10 @@ public String toPattern() {
      *
      * @param pattern String
      */
+    @SuppressWarnings({"array.access.unsafe.high","compound.assignment.type.incompatible"}) /*
+    #1, #2, #3, #4 - pos.getIndex() < pattern.length() => pos.getIndex() is a valid index for c
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/ExtendedMessageFormat.java]: `pos.getIndex()` returns different values at different points in this method, though! So the single check at the beginning is not sufficient to prove correctness.

I suggest trying a small refactoring: assign the value of `pos.getIndex()` to a new local before the start of the for-loop. Re-assign it after every call to `pos.next()` or another API which changes the position, and then use that variable throughout this method. That should eliminate warnings on the obviously correct parts of the code (like the `switch` statement), but will point us to the parts of the code that rely on another invariant.

File: src/main/java/org/apache/commons/lang3/text/ExtendedMessageFormat.java [hunk 5]
```diff
@@ -324,16 +332,20 @@ private Format getFormat(final String desc) {
      * @param pos current parse position
      * @return argument index
      */
+    @SuppressWarnings("argument.type.incompatible")/*
+    #1, #2 - pos.getIndex() < pattern.length() (condition for the loop) => pos.Index is a valid index for pattern
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/ExtendedMessageFormat.java]: Similar suggestion to my suggestion above about adding a local variable here.

File: src/main/java/org/apache/commons/lang3/text/ExtendedMessageFormat.java [hunk 6]
```diff
@@ -366,20 +378,23 @@ private int readArgumentIndex(final String pattern, final ParsePosition pos) {
      * @param pos current parse position
      * @return Format description String
      */
+    @SuppressWarnings("argument.type.incompatible")/*
+    #1, #2 - pos.getIndex() < pattern.length() as checked by the condition in the if statement
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/ExtendedMessageFormat.java]: Refactor to local if possible

File: src/main/java/org/apache/commons/lang3/text/ExtendedMessageFormat.java [hunk 7]
```diff
@@ -409,6 +424,7 @@ private String insertFormats(final String pattern, final ArrayList<String> custo
         int fe = -1;
         int depth = 0;
         while (pos.getIndex() < pattern.length()) {
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/ExtendedMessageFormat.java]: refactor to local

File: src/main/java/org/apache/commons/lang3/text/StrTokenizer.java [hunk 8]
```diff
@@ -656,7 +666,7 @@ private void checkTokenized() {
     /**
      * Adds a token to a list, paying attention to the parameters we've set.
      *
-     * @param list  the list to add to
+     * @param lisStrTokenizer.javat  the list to add to
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/StrTokenizer.java]: ?

File: src/main/java/org/apache/commons/lang3/text/StrTokenizer.java [hunk 9]
```diff
@@ -682,36 +692,40 @@ private void addToken(final List<String> list, String tok) {
      * @return the starting position of the next field (the character
      *  immediately after the delimiter), or -1 if end of string found
      */
-    private int readNextToken(final char[] srcChars, int start, final int len, final StrBuilder workArea, final List<String> tokenList) {
+    @SuppressWarnings("argument.type.incompatible") /*
+    #1, #2, #3, #4 - start < len as it is the condition for the while loop
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/StrTokenizer.java]: An `@LessThan` type should be inferred, then. What's the expected type of the last argument to `isMatch`?

File: src/main/java/org/apache/commons/lang3/text/StrTokenizer.java [hunk 10]
```diff
@@ -828,9 +843,10 @@ private int readWithQuotes(final char[] srcChars, final int start, final int len
      * @param quoteLen  the length of the matched quote, 0 if no quoting
      * @return true if a quote is matched
      */
+    @SuppressWarnings({"array.access.unsafe.low","array.access.unsafe.high"}) // #1 If pos + i >= len, srcChars[pos + i] and srcChars[quoteStart + i] does not happen. Also, pos >= quoteStart as is a virtue when it is called. This function is called only in readWithQuotes() which is called only by readNextTokens(), which declares pos and quotestart to either be start + quoteLen and start respectively or start and 0 respectively, where start and quoteLen both are @NonNegative.
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/StrTokenizer.java]: What does this mean:

"Also, pos >= quoteStart as is a virtue when it is called"

File: src/main/java/org/apache/commons/lang3/text/WordUtils.java [hunk 11]
```diff
@@ -271,6 +274,11 @@ public static String wrap(final String str, final int wrapLength, final String n
      *               if blank string is provided a space character will be used
      * @return a line with newlines inserted, <code>null</code> if null input
      */
+    @SuppressWarnings({"assignment.type.incompatible","compound.assignment.type.incompatible"}) /*
+    #1 matcher is on the substring of str that starts with offset index, hence (offset + matcher.end())'s max value = offset + str.length() - 1 - offset = str.length() - 1
+    #2, #3 matcher is on the substring of str that starts with offset index, hence (offset + matcher.start())'s max value = offset + str.length() - 1 - offset = str.length() - 1
+    #4 matcher is on the substring of str that starts with offset + wrapLength index, hence (offset + wrapLength + matcher.start())'s max value = offset + wrapLength + str.length() - 1 - offset - wrapLength = str.length() - 1
+    */
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/WordUtils.java:282]: Is it possible to express these invariants with `offset` arguments to annotations?

File: src/main/java/org/apache/commons/lang3/text/WordUtils.java [hunk 12]
```diff
@@ -286,19 +294,19 @@ public static String wrap(final String str, int wrapLength, String newLineStr, f
         }
         final Pattern patternToWrapOn = Pattern.compile(wrapOn);
         final int inputLineLength = str.length();
-        int offset = 0;
+        @IndexOrHigh("str") int offset = 0;
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/WordUtils.java:298]: Shouldn't this have an upperbound type that includes an offset vs `wrapLength`, at least?

File: src/main/java/org/apache/commons/lang3/text/WordUtils.java [hunk 13]
```diff
@@ -271,6 +274,11 @@ public static String wrap(final String str, final int wrapLength, final String n
      *               if blank string is provided a space character will be used
      * @return a line with newlines inserted, <code>null</code> if null input
      */
+    @SuppressWarnings({"assignment.type.incompatible","compound.assignment.type.incompatible"}) /*
+    #1 matcher is on the substring of str that starts with offset index, hence (offset + matcher.end())'s max value = offset + str.length() - 1 - offset = str.length() - 1
+    #2, #3 matcher is on the substring of str that starts with offset index, hence (offset + matcher.start())'s max value = offset + str.length() - 1 - offset = str.length() - 1
+    #4 matcher is on the substring of str that starts with offset + wrapLength index, hence (offset + wrapLength + matcher.start())'s max value = offset + wrapLength + str.length() - 1 - offset - wrapLength = str.length() - 1
+    */
     public static String wrap(final String str, int wrapLength, String newLineStr, final boolean wrapLongWords, String wrapOn) {
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/WordUtils.java]: Why doesn't `wrapLength` have an annotation? Can it really take on any value?

File: src/main/java/org/apache/commons/lang3/text/translate/CharSequenceTranslator.java [hunk 14]
```diff
@@ -79,6 +81,10 @@ public final String translate(final CharSequence input) {
      * @param out Writer to translate the text to
      * @throws IOException if and only if the Writer produces an IOException
      */
+    @SuppressWarnings("argument.type.incompatible") /*
+    #1, #2 - argument to write() need not be @NonNegative
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/translate/CharSequenceTranslator.java]: If that's so, do you want to make a PR against the annotated JDK to fix the annotation there?

File: src/main/java/org/apache/commons/lang3/text/translate/CharSequenceTranslator.java [hunk 15]
```diff
@@ -79,6 +81,10 @@ public final String translate(final CharSequence input) {
...
+    #3 - consumed is the number of codepoints used to represent the character, hence, the calue of pos remains a valid index for input throughout the loop
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/translate/CharSequenceTranslator.java]: "calue" -> "value"

File: src/main/java/org/apache/commons/lang3/text/translate/CharSequenceTranslator.java [hunk 16]
```diff
@@ -121,7 +127,8 @@ public final void translate(final CharSequence input, final Writer out) throws I
      * @return CharSequenceTranslator merging this translator with the others
      */
     public final CharSequenceTranslator with(final CharSequenceTranslator... translators) {
-        final CharSequenceTranslator[] newArray = new CharSequenceTranslator[translators.length + 1];
+        @SuppressWarnings("assignment.type.incompatible") // translators.length + 1 has minimum value 1
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/translate/CharSequenceTranslator.java]: Why is this necessary? If this annotation isn't inferred, this is a bug in the Index Checker.

File: src/main/java/org/apache/commons/lang3/text/translate/JavaUnicodeEscaper.java [hunk 17]
```diff
@@ -110,7 +112,8 @@ public JavaUnicodeEscaper(final int below, final int above, final boolean betwee
      */
     @Override
     protected String toUtf16Escape(final int codepoint) {
-        final char[] surrogatePair = Character.toChars(codepoint);
+        @SuppressWarnings("assignment.type.incompatible") // array has @MinLen(2) if codepoint > 0xffff, which is checked when this function is called
+        final char @MinLen(2) [] surrogatePair = Character.toChars(codepoint);
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/translate/JavaUnicodeEscaper.java:118]: Is it possible to annotate `codepoint` with an `@IntRange` annotation that expresses that information?

That won't prevent the warning suppression, but it will require that the check happens at call-sites.

File: src/main/java/org/apache/commons/lang3/text/translate/LookupTranslator.java [hunk 18]
```diff
@@ -46,15 +53,19 @@
      *
      * @param lookup CharSequence[][] table of size [*][2]
      */
-    public LookupTranslator(final CharSequence[]... lookup) {
+    @SuppressWarnings({"enhancedfor.type.incompatible","argument.type.incompatible"}) /*
+    #1 CharSequence is of the type [*][2] (according to the documentation), hence Charsequence has length 2
+    #2 Minimum 1 row ensures seq[0].charAt(0) to be valid
```
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/translate/LookupTranslator.java]: Having at least one row doesn't mean that there is a `charAt` position 0.

**Review by kelloggm — CHANGES_REQUESTED** [review]
Overall this looks good - nice work. I've left some comments where I think things can be improved.

File: src/main/java/org/apache/commons/lang3/text/StrTokenizer.java (hunk 9 above)
Comment (Kakarot-SSJ4) [reply src/main/java/org/apache/commons/lang3/text/StrTokenizer.java]: It is for start only and not the last argument, the last argument has no annotation.
Also, I had counted the characters from the start where the error is being registered.

File: src/main/java/org/apache/commons/lang3/text/WordUtils.java (hunk 11 above)
Comment (Kakarot-SSJ4) [reply src/main/java/org/apache/commons/lang3/text/WordUtils.java:282]: No, because the maximum value of spaceToWrapAt is str.length() - 1 and it does not depend on offset.

File: src/main/java/org/apache/commons/lang3/text/WordUtils.java (hunk 12 above)
Comment (Kakarot-SSJ4) [reply src/main/java/org/apache/commons/lang3/text/WordUtils.java:298]: inputLineLength - offset <= wrapLength is checked in the loop.

File: src/main/java/org/apache/commons/lang3/text/translate/CharSequenceTranslator.java (hunk 16 above)
Comment (Kakarot-SSJ4) [reply src/main/java/org/apache/commons/lang3/text/translate/CharSequenceTranslator.java]: It isn't inferred and I think it is due to the imprecision in the Index Checker related to multi-dimensional arrays

File: src/main/java/org/apache/commons/lang3/text/translate/LookupTranslator.java (hunk 18 above)
Comment (Kakarot-SSJ4) [reply src/main/java/org/apache/commons/lang3/text/translate/LookupTranslator.java]: It has 2 columns for sure, hence minimum 1 row should ensure the existence of charAt position 0

File: src/main/java/org/apache/commons/lang3/text/WordUtils.java (hunk 11 above)
Comment (kelloggm) [reply src/main/java/org/apache/commons/lang3/text/WordUtils.java:282]: I think it would be helpful to have that explicitly in this comment, because it wasn't obvious to me when reading the explanation, and makes it much more clear.

**Comment by Kakarot-SSJ4** [comment]
Annotated WordUtils.java and StrSubstitutor.java
